lib/portage/dbapi/bintree.py

```diff
@@ -1916,6 +1929,56 @@ def _read_metadata(self, filename, st, keys=None, binpkg_format=None):
 
         return metadata
 
+    def _inject_repo_revisions(self, header, repo_revisions):
+        """
+        Inject REPO_REVISIONS from a package into the index header,
+        using a history of synced revisions to guarantee forward
+        progress. This queries the relevant repos to check if any
+        new revisions have appeared in the absence of a proper sync
+        operation.
+
+        This does not expose REPO_REVISIONS that do not appear in
+        the sync history, since such revisions suggest that the
+        package was not built locally, and in this case its
+        REPO_REVISIONS are not intended to be exposed.
+        """
+        synced_repo_revisions = get_repo_revision_history(
+            self.settings["EROOT"],
+            [self.settings.repositories[repo_name] for repo_name in repo_revisions],
+        )
+        header_repo_revisions = (
+            json.loads(header["REPO_REVISIONS"]) if header.get("REPO_REVISIONS") else {}
+        )
+        for repo_name, repo_revision in repo_revisions.items():
+            rev_list = synced_repo_revisions.get(repo_name, [])
+            header_rev = header_repo_revisions.get(repo_name)
+            if not rev_list or header_rev in (repo_revision, rev_list[0]):
+                continue
+            try:
+                header_rev_index = (
+                    None if header_rev is None else rev_list.index(header_rev)
+                )
+            except ValueError:
+                header_rev_index = None
+            try:
+                repo_revision_index = rev_list.index(repo_revision)
+            except ValueError:
+                repo_revision_index = None
+            if repo_revision_index is not None and (
+                header_rev_index is None or repo_revision_index < header_rev_index
+            ):
+                # There is forward progress when repo_revision is more recent
+                # than header_rev or header_rev was not found in the history.
```

Comment:
The repo_revisions history file exists solely so that we can perform this comparison in order to ensure forward progress.

As an alternative that avoids the need for a repo revision history, we could simply expose the latest synced revision here regardless of the revision that the package was built with. In most cases I would expect the package would be built against the latest revision anyway (unless there's a cron job that syncs periodically in the background).

Reply:
@akhuettel, for gentoo binhost builds, is there a cron job that sync repos in the background or is sync triggered prior to builds? Either will work if we track the history of repo revisions, but I'm just checking since tracking repo revision history just so we can validate revisions exposed in the binhost Packages header could be overkill.

Reply:
@zmedico The sync is triggered and completed before the build starts (both for the dedicated binhosts and for the catalyst builds).
The dedicated binhosts use a separate (from the host) repo dir for that, and the catalyst builds have their squashfs snapshot.